### DISCUSSION TITLE
I can´t add a column to a certain layer in Qgis for a plugin using python

### DISCUSSION THREAD

**Discussion by TheJoseph24** [community/community#86060] [Programming Help]
### Body

Hi!
I´m currently developing a plugin for Qgis using Qt Designer and python, and one of the things I´m trying to do is to calculate the length of all the lines of a layer and add the values to a new column in the attribute field. I´ve done this once to calculate the area for a different layer and it worked, but when I readjust the code to give the length values, the field created still goes to the first layer attribute table instead of the one I want (as you can see by the print screen).
To give some more context, I´ll drop the code in a PDF format and the snippet of the code I used to creat a new field and add it to the attribute field. Can someone help me?

[Code.pdf](https://github.com/community/community/files/13877224/Code.pdf)
![Problme](https://github.com/community/community/assets/111004787/eb1a3286-137d-4357-bbfc-45b53bcc8019)
lines_edit=QgsVectorLayer(str(output),"closest_points","ogr")
     if lines_edit.isValid():
            idx_length = lines_edit.fields().lookupField("Comprimento")
            if idx_length == -1:
                lines_edit.dataProvider().addAttributes([QgsField("Comprimento", QVariant.Double)])
                lines_edit.updateFields()
            lines_edit.startEditing()
            for feature2 in lines_edit.getFeatures():
                length_calc=feature.geometry().length()
                feature2.setAttribute("Comprimento", length_calc)
                lines_edit.updateFeature(feature)
            lines_edit.commitChanges()
            QMessageBox.about(self.dlg, "Sucesso:","o comprimento das linhas é"+str(length_calc))
        else:
            QMessageBox.about(self.dlg,"Aviso","O campo não é válido")

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by kmadaminjanova0508** [comment]
Hi @TheJoseph24

I think there might be an issue in your code related to the iteration over features and updating the attributes. 

 
lines_edit = QgsVectorLayer(str(output), "closest_points", "ogr")

if lines_edit.isValid():
    idx_length = lines_edit.fields().lookupField("Comprimento")

    if idx_length == -1:
        lines_edit.dataProvider().addAttributes([QgsField("Comprimento", QVariant.Double)])
        lines_edit.updateFields()

    lines_edit.startEditing()

    for feature2 in lines_edit.getFeatures():
        # Ensure the 'feature' variable is correctly defined. Use 'feature2' for the iteration.
        length_calc = feature2.geometry().length()
        feature2.setAttribute("Comprimento", length_calc)
        lines_edit.updateFeature(feature2)

    lines_edit.commitChanges()

    QMessageBox.about(self.dlg, "Sucesso:", "O comprimento das linhas foi calculado com sucesso.")
else:
    QMessageBox.about(self.dlg, "Aviso", "A camada não é válida.")


here I replaced the occurrences of feature with feature2 within the loop, as that seems to be the variable used for iteration over features and I  also added a check to make sure the layer is in editing mode (lines_edit.startEditing()) before making changes and committed the changes after the loop (lines_edit.commitChanges()).The length_calc is now calculated based on feature2.geometry().length() instead of feature.geometry().length().Please also Check that you have the correct variable names in place and that you are iterating over the correct layer (lines_edit). Also, make sure that the layer is in editing mode when making changes to its features.This should help you correctly calculate the length of the lines and update the attribute field ;)

**Reply by TheJoseph24** [reply]
Hi, I tried everything you've told me to do and unfortunately, it still gave me the same result, wich is weird because I don´t reference the layer the field goes to a single time in that code snippet. Still, thank you for trying to help.
